ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,327 @@
+
+## State on Followers
+
+Instead of running the same processing state machine on the followers, we replay the events from the replicated log. For that we have a ReplayStateMachine, which allows to continuously replay events and apply them to the state. We call it replay, since they have been already applied on the leader side and have been produced by him. Furthermore, we will reuse this state machine on bootstrap a leader partition such that the term "replay" makes sense here as well.
+
+Raft ROLEs are transient states, which means it is likely that a Leader change happens. How the system reacts on these role changes can be seen in the following picture. With the following process, we minimize the time for the Follower-to-Leader transition, which impacts the process execution latency.
+
+![stateOnFollower](images/stateOnFollower.png)
+
+On bootstrapping of a Zeebe Partition we first install all needed services, like the StreamProcessor, the SnapshotDirector etc. In later images we will collapse the "Install Zeebe Partition" as call activity, to make it not so overwhelming. After installing our services we go over to a state we call "Stream Processing", which is our real business logic and the heart of each partition. Based on the RAFT role we perform different actions on the "Stream Processing". We can distinguish four cases which we will explain more in depth in the following sections:
+
+  * Bootstrapping Leader Partition, no already running partition
+  * Bootstrapping Follower Partition, no already running partition
+  * Switch to Leader Partition, from an already running Follower partition
+  * Switch to Follower Partition, from an already running Leader Partition.
+
+In the following section we will use the term processing and replay heavily, please read the [ZEP004](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md) if you want to know more about it.
+
+### Bootstrap Leader Partition
+
+![stateOnFollower-BootstrapLeader](images/stateOnFollower-BootstrapLeader.png)
+
+In this scenario we have no already running Zeebe partition and installed it completely new. During installing our services we restore the state, via copying the latest snapshot into the runtime folder. After installing all services, we start with the "Stream Processing", which is on the Leader separated in two parts. The first part is the processing, which consist of the following steps:
+
+  * **Replay until end**, here we replay all existing events on the Log, which were not part of the Snapshot. This is necessary to get the latest state, before processing.
+  * **Reset Dispatcher Position**, here we set the Dispatcher position which generates the position for our follow-up events.
```

Comment:
I guess we generally want to reset the dispatcher entirely, no? I'd be a bit worried if it's long living of having left over stuff in there.

Reply:
Yes we should wipe it right.